File: Baekjoon/DFS/9663.py
# https://www.acmicpc.net/problem/9663

import sys
input = sys.stdin.readline

n = int(input())
graph = [0] * n
res = 0

def check(x):
    for i in range(x):
        if graph[x] == graph[i] or abs(x - i) == abs(graph[x] - graph[i]):
            return False
    return True

def dfs(x):
    global res

    if x == n:
        res += 1
        return
    
    for i in range(n):
        graph[x] = i
        if check(x):
            dfs(x + 1)

dfs(0)
print(res)

File: Baekjoon/DFS/9663.cpp
// https://www.acmicpc.net/problem/9663

#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

int N;
int cnt = 0;
vector<int> col;

bool isPossible(int level) {
    for (int i = 0; i < level; i++) {
        if (col[i] == col[level] || abs(col[level] - col[i]) == level - i) {
            return false;
        }
    }
    return true;
}

void dfs(int level) {
    if (level == N) {
        cnt++;
        return;
    }
    for (int i = 0; i < N; i++) {
        col[level] = i;
        if (isPossible(level)) {
            dfs(level + 1);
        }
    }
}

int main() {
    cin >> N;
    col.resize(N);
    dfs(0);
    cout << cnt << endl;
    return 0;
}